lib/src/Pos/Util/UserPublic.hs

```diff
@@ -0,0 +1,284 @@
+lockFilePath = (<> ".lock")
+
+-- | Checks if this user public instance can be dumped back to
+-- file. If not, using 'writeUserPublic' and 'writeUserPublicRelease'
+-- will result in error.
+canWrite :: UserPublic -> Bool
+canWrite up = up ^. upLock . to isJust
+
+instance Default UserPublic where
+    def = UserPublic [] Nothing "" Nothing
+
+-- | It's not network/system-related, so instance shouldn't be under
+-- @Pos.Binary.*@.
+instance Bi UserPublic where
+  encode us = encodeListLen 2 <> encode (_upKeys us)
+                              <> encode (_upWallet us)
+  decode = do
+    enforceSize "UserPublic" 2
+    pKeys  <- decode
+    wallet <- decode
+    return $ def
+        & upKeys   .~ pKeys
+        & upWallet .~ wallet
+
+#ifdef POSIX
+-- | Constant that defines file mode 600 (readable & writable only by owner).
+mode600 :: PSX.FileMode
+mode600 = PSX.unionFileModes PSX.ownerReadMode PSX.ownerWriteMode
+
+-- | Return only the access part of the file mode (like owner:rw-, etc).
+getAccessMode :: (MonadIO m) => FilePath -> m PSX.FileMode
+getAccessMode path = do
+    mode <- liftIO $ PSX.fileMode <$> PSX.getFileStatus path
+    return $ PSX.intersectFileModes mode PSX.accessModes
+
+-- | Set mode 600 on a given file, regardless of its current mode.
+setMode600 :: (MonadIO m) => FilePath -> m ()
+setMode600 path = liftIO $ PSX.setFileMode path mode600
+#endif
+
+#ifdef POSIX
+ensureModeIs600 :: (MonadIO m, WithLogger m) => FilePath -> m ()
+ensureModeIs600 path = do
+    accessMode <- getAccessMode path
+    unless (accessMode == mode600) $ do
+        logWarning $
+            sformat ("Key file at "%build%" has access mode "%oct%" instead of 600. Fixing it automatically.")
+            path accessMode
+        setMode600 path
+#endif
+
+-- | Create user public file at the given path, but only when one doesn't
+-- already exist.
+#ifdef POSIX
+initializeUserPublic :: (MonadIO m, WithLogger m) => FilePath -> m ()
+#else
+initializeUserPublic :: (MonadIO m) => FilePath -> m ()
+#endif
+initializeUserPublic publicPath = do
+    exists <- liftIO $ doesFileExist publicPath
+#ifdef POSIX
+    if exists
+    then ensureModeIs600 publicPath
+    else do
+        createEmptyFile publicPath
+        setMode600 publicPath
+#else
+    unless exists $ createEmptyFile publicPath
+#endif
+  where
+    createEmptyFile :: (MonadIO m) => FilePath -> m ()
+    createEmptyFile = liftIO . flip writeFile mempty
+
+-- | Reads user public from file, assuming that file exists,
+-- and has mode 600, throws exception in other case
+#ifdef POSIX
+readUserPublic :: (MonadIO m, WithLogger m) => FilePath -> m UserPublic
+#else
+readUserPublic :: (MonadIO m) => FilePath -> m UserPublic
+#endif
+readUserPublic path = do
+#ifdef POSIX
+    ensureModeIs600 path
+#endif
+    takeReadLock path $ do
+        content <- either (throwM . UserPublicDecodingError . toText) pure .
+                   decodeFull' =<< BS.readFile path
+        pure $ content & upPath .~ path
```

Comment:
Just being a little philosophical here.

There is actually more than one kind of exception that can happen here;
* An `IOException` if `path` does not exist
* the user defined `UserPublicDecodingError` on failure to decode
and there is nothing in the type to indicate either. An alternative (ignoring the non-POSIX case) to this would be:

```
import Control.Exception (IOError, handle)

data UserPublicError
    = UserPublicDecodingError !Text
    | UserPublicIOError !IOError

readUserPublic :: FilePath -> IO (Either UserPublicError UserPublic)
readUserPublic path = do
    handle handler $ do
        ensureModeIs600 path
        takeReadLock path $ do
            content <- either (UserPublicDecodingError . toText) pure .
                           decodeFull' =<< BS.readFile path
            pure $ content & upPath .~ path
  where
    handler :: IOError -> IO (Either UserPublicError a) 
    handler = pure . UserPublicIOError . toText . show
```

The reason I like this is because:
* It doesn't ignore the possibility of the file not being found
* It *forces* the caller to deal with that possibility and makes it more difficult for them to ignore it (ie catching the exception and silently dropping it).
* It treats both unhappy cases (the possible exceptions) as equals to the happy case (file found and decode successful).

Reply:
I like @erikd's suggestion. I would change the error handling logic to check the `IOError` was actually a FileNotFound or whatever it throws in that case, and only rewrap the necessary information.